odl/space/space_utils.py

```diff
@@ -281,6 +284,303 @@ def rn(shape, dtype=None, impl='numpy', **kwargs):
+                        'returning `self` in an `adjoint` property using '
+                        '`auto_weighting` is not allowed')
+
+                # This is for cached adjoints: don't double-wrap
+                if hasattr(adj, '_call_unweighted'):
+                    return adj
+                else:
+                    return auto_weighting._instance_wrapper(adj, optimize)
+
+            return weighted_adjoint
+
+        else:
+            raise TypeError(
+                "`auto_weighting` can only be applied to 'adjoint' methods "
+                '(@auto_weighting decorator)')
+
+    @staticmethod
+    def _instance_wrapper(unweighted_adjoint, optimize=True):
+        """Wrapper for `Operator` instances."""
+        # Use notions of the original operator, not the adjoint
+        dom_weighting = unweighted_adjoint.range.weighting
+        ran_weighting = unweighted_adjoint.domain.weighting
+
+        if isinstance(dom_weighting, ArrayWeighting):
+            dom_w_type = 'array'
+            dom_w = dom_weighting.array
+        elif isinstance(dom_weighting, ConstWeighting):
+            dom_w_type = 'const'
+            dom_w = dom_weighting.const
+        else:
+            raise TypeError(
+                'weighting of `unweighted_adjoint.range` must be of '
+                'type `ArrayWeighting` or `ConstWeighting`, got {}'
+                ''.format(type(dom_weighting)))
+
+        if isinstance(ran_weighting, ArrayWeighting):
+            ran_w_type = 'array'
+            ran_w = ran_weighting.array
+        elif isinstance(ran_weighting, ConstWeighting):
+            ran_w_type = 'const'
+            ran_w = ran_weighting.const
+        else:
+            raise TypeError(
+                'weighting of `unweighted_adjoint.domain` must be of '
+                'type `ArrayWeighting` or `ConstWeighting`, got {}'
+                ''.format(type(ran_weighting)))
+
+        # Compute the effective weights and mark constants 1.0 as to be
+        # skipped
+        if not optimize:
+            new_dom_w, new_ran_w = dom_w, ran_w
+            skip_dom = (dom_w_type == 'const' and dom_w == 1.0)
+            skip_ran = (ran_w_type == 'const' and ran_w == 1.0)
+        elif dom_w_type == 'array' and ran_w_type == 'array':
+            new_dom_w, new_ran_w = dom_w, ran_w
+            skip_dom = skip_ran = False
+        elif dom_w_type == 'array' and ran_w_type == 'const':
+            new_dom_w = dom_w / ran_w
+            new_ran_w = 1.0
+            skip_dom = False
+            skip_ran = True
+        elif dom_w_type == 'const' and ran_w_type == 'array':
+            new_dom_w = 1.0
+            new_ran_w = ran_w / dom_w
+            skip_dom = True
+            skip_ran = False
+        elif dom_w_type == 'const' and ran_w_type == 'const':
+            if unweighted_adjoint.domain.size < unweighted_adjoint.range.size:
+                new_dom_w = 1.0
+                new_ran_w = ran_w / dom_w
```

Comment:
Are we sure this always makes sense? What about operators `C -> R` etc?

Reply:
Yes, because weightings can never be complex. (Note to self: make sure that this is checked.)